core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -355,26 +398,19 @@ class ReplicaFetcherThread(name: String,
 
           // if we get any unexpected exception, mark all partitions with an error
           result = partitions.map { case (tp, _) =>
-            tp -> new EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH_OFFSET)
+            tp -> new EpochEndOffset(Errors.forException(t), UNDEFINED_EPOCH, UNDEFINED_EPOCH_OFFSET)
```

Comment:
I have a question regarding this. In OffsetsForLeaderEpochResponse, I used `partitionStruct.setIfExists` and `partitionAndEpoch.getOrElse` for new leader epoch field vs. explicitly checking version for setting/getting fields. I think, it should be correct to e.g., return the leader epoch but make sure to use the older protocol in fetcher if the inter-broker protocol is below 2.0. Or is it preferred to check the version and not set the field in OffsetsForLeaderEpochResponse? I looked at the code, and it looks like different request/response use different approaches (either one I described).

This particular line of code is correct independent of version, since undefined offsets means falling back to high watermark, independent if epoch is set or not (or defined or not defined). 

I agree I need to update the truncation logic to make sure the behavior is consistent with the protocol version.

Reply:
I looked at OffsetsForLeaderEpochRequest.Builder, and I see that we pass version to the builder:
`public OffsetsForLeaderEpochRequest build(short version) {`
` public OffsetsForLeaderEpochRequest(Struct struct, short version) {`
` @Override
    protected Struct toStruct() {
        Struct requestStruct = new Struct(ApiKeys.OFFSET_FOR_LEADER_EPOCH.requestSchema(version()));
... `
Or am I missing something?

Reply:
Oh I see now, version is not in OffsetForLeaderEpochResponse constructor, not request. Will fix.